### ISSUE TITLE
Extreme C-S latencies and timeouts after a node was forcibly removed from the cluster

### ISSUE THREAD

**Issue by ShlomiBalalis** [scylladb/scylla#7580]
*Installation details*
Scylla version (or git commit hash): 4.3.rc0-0.20201028.bbef05ae3c with build-id adad9e1dae9eb9dbebab38bd5e7258d5c7360350
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-01cc969208fae7a3a (eu-west-1)
Scenario: **longevity-mv-si-4days**

At `2020-11-09 08:46:30.762` a `TerminateAndRemoveNodeMonkey` nemesis has started, with node#3 as the target node.

The stages of the nemesis are as such: First, SCT stops scylla at the target node, and then terminates it. Afterwards we execute a `nodetool repair` on each of the other nodes. At the end, we execute `nodetool removenode` on the target node, and init a new node.

The repairs we execute on each of the healthy nodes, however, all fail from seemingly the same issue:
```
2020-11-09T08:51:04+00:00  longevity-mv-si-4d-4-3-db-node-d0cc0297-1 !WARNING | scylla: [shard 13] repair - Repair 800 out of 801 ranges, id=[id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980], shard=13, keyspace=mview, table={users_by_first_name, users_by_email, users_by_address, users, users_by_password, users_by_initials, users_by_last_name}, range=(9159346189907813475, 9184426635097401780], peers={10.0.3.22, 10.0.3.139}, live_peers={10.0.3.139}, status=partial
2020-11-09T08:51:08+00:00  longevity-mv-si-4d-4-3-db-node-d0cc0297-1 !WARNING | scylla: [shard 13] repair - repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 13 failed - 378 ranges failed
2020-11-09T08:51:08+00:00  longevity-mv-si-4d-4-3-db-node-d0cc0297-1 !WARNING | scylla: [shard 13] repair - repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 13 failed: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 13 failed to repair 378 sub ranges)
2020-11-09T08:51:08+00:00  longevity-mv-si-4d-4-3-db-node-d0cc0297-1 !WARNING | scylla: [shard 0] repair - repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] failed: std::runtime_error ({shard 0: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 0 failed to repair 378 sub ranges), shard 1: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 1 failed to repair 378 sub ranges), shard 2: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 2 failed to repair 378 sub ranges), shard 3: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 3 failed to repair 378 sub ranges), shard 4: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 4 failed to repair 378 sub ranges), shard 5: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 5 failed to repair 378 sub ranges), shard 6: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 6 failed to repair 378 sub ranges), shard 7: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 7 failed to repair 378 sub ranges), shard 8: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 8 failed to repair 378 sub ranges), shard 9: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 9 failed to repair 378 sub ranges), shard 10: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 10 failed to repair 378 sub ranges), shard 11: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 11 failed to repair 378 sub ranges), shard 12: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 12 failed to repair 378 sub ranges), shard 13: std::runtime_error (repair id [id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980] on shard 13 failed to repair 378 sub ranges)})
```

![Screenshot from 2020-11-10 13-39-47](https://user-images.githubusercontent.com/40186893/98669580-3d746880-235a-11eb-9270-b6e459eb8696.png)

![Screenshot from 2020-11-10 13-36-32](https://user-images.githubusercontent.com/40186893/98669528-29c90200-235a-11eb-876e-7b9921deaac2.png)

The repairs run between `2020-11-09 08:48:18,324` and `2020-11-09 08:57:16,816`, and in that time the C-S latency become extremely high (to the point of several seconds) and as a result the number of request the cluster is able to handle nearly reaches zero, as you can see in the images above

Moreover, when the removenode command is executed, between `2020-11-09 08:57:16,816` and `2020-11-09 09:10:37,841`, the latency reaches new heights and a large amount of read and write timeouts occur:

![Screenshot from 2020-11-10 13-47-05](https://user-images.githubusercontent.com/40186893/98670466-6a754b00-235b-11eb-9b16-0de628a368e6.png)

Only when the removenode command ends, and a new node is added to the cluster, the cluster start to function properly again.

From C-S log:
cassandra-stress-l0-c0-k1-8d1ac641-2109-4fbd-9340-e8019ea0a594.log
```
08:57:16.154 [cluster1-nio-worker-4] DEBUG com.datastax.driver.core.Connection - Connection[/10.0.3.92:9042-15, inFlight=0, closed=false] Response received on stream 8320 but no handler set anymore (either the request has timed out or it was closed due to another error). Received message is ERROR READ_TIMEOUT: Operation timed out for sec_index.users - received only 1 responses from 2 CL=QUORUM.
```
cassandra-stress-l0-c0-k1-e5718ecb-49e3-4464-937a-de83b41bc46b.log
```
19:05:58.812 [cluster1-nio-worker-5] DEBUG com.datastax.driver.core.Connection - Connection[/10.0.3.22:9042-4, inFlight=10, closed=false] Response received on stream 8448 but no handler set anymore (either the request has timed out or it was closed due to another error). Received message is ERROR WRITE_TIMEOUT: Operation timed out for mview.users - received only 1 responses from 2 CL=QUORUM.
```

Logs:
```
Being collected
```
* ATM, the logs can be viewed at the SCT runner of the job, at `3.250.68.178`.

Node list:
```
longevity-mv-si-4d-4-3-db-node-d0cc0297-1 [34.243.32.181 | 10.0.0.120]
longevity-mv-si-4d-4-3-db-node-d0cc0297-2 [54.217.23.112 | 10.0.2.17]
longevity-mv-si-4d-4-3-db-node-d0cc0297-3 [54.75.88.71 | 10.0.3.22]
longevity-mv-si-4d-4-3-db-node-d0cc0297-4 [63.35.164.214 | 10.0.1.195]
longevity-mv-si-4d-4-3-db-node-d0cc0297-5 [52.30.10.134 | 10.0.3.45]
longevity-mv-si-4d-4-3-db-node-d0cc0297-6 [3.249.245.31 | 10.0.3.227]
longevity-mv-si-4d-4-3-db-node-d0cc0297-7 [3.250.107.212 | 10.0.3.118]
longevity-mv-si-4d-4-3-db-node-d0cc0297-8 [34.251.206.126 | 10.0.0.25]
longevity-mv-si-4d-4-3-db-node-d0cc0297-9 [34.249.28.135 | 10.0.0.142]
longevity-mv-si-4d-4-3-db-node-d0cc0297-10 [52.16.132.34 | 10.0.3.100]
longevity-mv-si-4d-4-3-db-node-d0cc0297-11 [54.246.44.87 | 10.0.3.159]
longevity-mv-si-4d-4-3-db-node-d0cc0297-12 [34.246.178.39 | 10.0.3.139] 
longevity-mv-si-4d-4-3-db-node-d0cc0297-13 [52.214.232.196 | 10.0.2.163]
longevity-mv-si-4d-4-3-db-node-d0cc0297-14 [54.246.177.225 | 10.0.3.92] 
longevity-mv-si-4d-4-3-db-node-d0cc0297-15 [54.155.94.122 | 10.0.0.70]
longevity-mv-si-4d-4-3-db-node-d0cc0297-16 [34.253.86.58 | 10.0.3.26]
```

Live monitor: http://54.171.83.95:3000/d/iawXcE2Mz/scylla-per-server-metrics-nemesis-master?orgId=1&from=1604911188919&to=1604916876378

**Comment by slivne** [comment]
@asias https://docs.scylladb.com/operating-scylla/procedures/cluster-management/remove_node/ - can we run repair when one of the replicas is down ? how can we solve this.

**Comment by slivne** [comment]
Please note we know there is an issue with c-s and how it handles the case of a node being killed - the period here seems much longer

**Comment by slivne** [comment]
So we did not run in SCT in 4.2 remvenode - lets try to figure out whats going on

**Comment by asias** [comment]
> @asias https://docs.scylladb.com/operating-scylla/procedures/cluster-management/remove_node/ - can we run repair when one of the replicas is down ? how can we solve this.

We can. If a node is down, repair continues in best effort mode.

**Comment by slivne** [comment]
@asias does it mean there is a special flag for this - how do we ask repair to run in best effort mode ?

**Comment by slivne** [comment]
@asias so items that need to cleared:
1. how to run repair in best effort so it does not end in failure ?
2. why durring the repair the trafiic drops 
3. why in removenode (after repair there is no traffic at all)

**Comment by asias** [comment]
> @asias does it mean there is a special flag for this - how do we ask repair to run in best effort mode ?
> 
> For QA all the repairs failed - check the logs above

Repair will report the repair job failed if it is in best effort mode (when repair followers are down).

**Comment by asias** [comment]
> @asias so items that need to cleared:
> 
>     1. how to run repair in best effort so it does not end in failure ?

We do not support  repair option to allow ignore some of the nodes.

> 
>     2. why durring the repair the trafiic drops

Did you mean why c-s QPS dropped?

> 
>     3. why in removenode (after repair there is no traffic at all)

Did you mean why c-s QPS dropped to zero?

**Comment by asias** [comment]
I did not find the scylla logs for the run.

**Comment by asias** [comment]
There was no repair io during repair (08:47 to 08:57) and removenode (08:57 09:10)
![image](https://user-images.githubusercontent.com/320076/99252391-d5d28780-2849-11eb-8534-0898107c313f.png)

During repair (08:47 to 08:57), the cpu load goes to 100%.  

![image](https://user-images.githubusercontent.com/320076/99252625-36fa5b00-284a-11eb-8530-e5bcaceb4cdc.png)

**Comment by roydahan** [comment]
@asias the cluster is still alive.
I think we lost the logs of SCT.

**Comment by asias** [comment]
Btw, I have no idea how to access the SCT runner and check the logs: `ATM, the logs can be viewed at the SCT runner of the job, at 3.250.68.178.`

If I understand correctly, SCT can  provide a link to the logs only AFTER the job is done. 

@roydahan  May I suggest to add a new feature to SCT so that  it can collect the logs at the point in time even if it is still running and provide a link to the logs and metric data.

**Comment by asias** [comment]
I can confirm from the log that node 3 10.0.3.22 was shutdown. The repair failed because the node3 was down. It was a partial repair so the nodetool repair returned fail instead of success. This is expected. 
```
Nov 09 08:48:25 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 10] repair - Repair 2 out of 801 ranges, id=[id=2568, uuid=4bc908e6-d6d1-40c8-8301-b9065cb7b980], shard=10, keyspace=mview, table={users_by_first_name, users_by_email, users_by_address, users, users_by_password, users_by_initials, users_by_last_name}, range=(-9217149796026851667, -9216145495964934463], peers={10.0.2.163, 10.0.3.22}, live_peers={10.0.2.163}, status=partial

```

```
Nov 09 08:47:00 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] gossip - InetAddress 10.0.3.22 is now DOWN, status = shutdown
Nov 09 08:58:44 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] gossip - InetAddress 10.0.3.22 is now DOWN, status = removing
Nov 09 09:10:37 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] gossip - InetAddress 10.0.3.22 is now DOWN, status = removed
Nov 09 09:10:37 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] gossip - Node 10.0.3.22 will be removed from gossip at [2020-11-12 09:10:35]: (expire = 1605172235704, now = 1604913037409, diff = 259198 seconds)
```

**Comment by asias** [comment]
MV Update failure during 10.0.3.139 is being replaced. 

The log was flushed with such repeated messages below. 

[scyllaadm@longevity-mv-si-4d-4-3-db-node-d0cc0297-1 ~]$ cat 2.mini |grep 'view - Error applying view update to 10.0.3.139'|wc -l
20743

@psarna @nyh why the futures are ignored in MV?


```
Nov 09 13:59:53 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] gossip - InetAddress 10.0.3.139 is now DOWN, status = shutdown

Nov 09 14:45:28 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] gossip - InetAddress 10.0.3.139 is now UP, status = UNKNOWN

Nov 09 14:45:28 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] storage_proxy - Exception when communicating with 10.0.3.139, to read from mview.users_by_address: std::runtime_error (Can't find a column family with UUID ab84d150-21a9-11eb-94ae-000000000003)
Nov 09 14:46:50 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] storage_service - Node 10.0.3.139 is replacing existing node 10.0.3.139 with host_id=5aa711b5-2b3a-4451-bbd6-044fe47def1d, existing_tokens=

Nov 09 14:46:50 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] token_metadata - Added node 10.0.3.139 as pending replacing endpoint which replaces existing node 10.0.3.139


Nov 09 14:46:50 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] view - Error applying view update to 10.0.3.139: seastar::broken_promise (broken promise)
Nov 09 14:46:50 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] seastar - Exceptional future ignored: seastar::broken_promise (broken promise), backtrace:    0x2fb794d
                                                                           0x2fb7ba0
                                                                           0x2fb8009
                                                                           0x2b03617
                                                                           0x2b0375c
                                                                           0x1d4fef9
                                                                           0x2b593a7
                                                                           0x2b5965e
                                                                           0x2b98f45
                                                                           0x2aecfce
                                                                           0x2aeda5e
                                                                           0xd4c992
                                                                           /opt/scylladb/libreloc/libc.so.6+0x27041
                                                                           0xc2d16d
                                                                           --------
                                                                           seastar::continuation<seastar::internal::promise_base_with_type<void>, db::view::mutate_MV(dht::token const&, std::vector<frozen_mutation_and_schema, std::allocator<frozen_mutation_and_schema> >, db::view::stats&, cf_stats&, tracing::trace_state_ptr, seastar::semaphore_units<seastar::default_timeout_exception_factory, seastar::lowres_clock>, seastar::bool_class<service::allow_hints_tag>, seastar::bool_class<db::view::wait_for_all_updates_tag>)::{lambda(seastar::future<void>&&)#3}, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, {lambda(seastar::future<void>&&)#3}>({lambda(seastar::future<void>&&)#3}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, {lambda(seastar::future<void>&&)#3}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
Nov 09 14:46:50 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] view - Error applying view update to 10.0.3.139: seastar::broken_promise (broken promise)
Nov 09 14:46:50 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 0] seastar - Exceptional future ignored: seastar::broken_promise (broken promise), backtrace:    0x2fb794d
                                                                           0x2fb7ba0
                                                                           0x2fb8009
                                                                           0x2b03617
                                                                           0x2b0375c
                                                                           0x1d4fef9
                                                                           0x2b593a7
                                                                           0x2b5965e
                                                                           0x2b98f45
                                                                           0x2aecfce
                                                                           0x2aeda5e
                                                                           0xd4c992
                                                                           /opt/scylladb/libreloc/libc.so.6+0x27041
                                                                           0xc2d16d
                                                                           --------
                                                                           seastar::continuation<seastar::internal::promise_base_with_type<void>, db::view::mutate_MV(dht::token const&, std::vector<frozen_mutation_and_schema, std::allocator<frozen_mutation_and_schema> >, db::view::stats&, cf_stats&, tracing::trace_state_ptr, seastar::semaphore_units<seastar::default_timeout_exception_factory, seastar::lowres_clock>, seastar::bool_class<service::allow_hints_tag>, seastar::bool_class<db::view::wait_for_all_updates_tag>)::{lambda(seastar::future<void>&&)#3}, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, {lambda(seastar::future<void>&&)#3}>({lambda(seastar::future<void>&&)#3}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, {lambda(seastar::future<void>&&)#3}&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
```

**Comment by asias** [comment]
semaphore_timed_out / smp_service_group in stroage_proxy: 

```
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 3] storage_proxy - Failed to apply mutation from 10.0.3.92#3: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 3] storage_proxy - Failed to apply mutation from 10.0.3.92#3: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 3] storage_proxy - Failed to apply mutation from 10.0.3.139#3: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 11] storage_proxy - Failed to apply mutation from 10.0.3.139#11: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 2] storage_proxy - Failed to apply mutation from 10.0.3.92#2: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->2 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 2] storage_proxy - Failed to apply mutation from 10.0.3.92#2: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->2 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 3] storage_proxy - Failed to apply mutation from 10.0.3.139#3: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 2] storage_proxy - Failed to apply mutation from 10.0.3.92#2: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->2 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 2] storage_proxy - Failed to apply mutation from 10.0.3.139#2: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->2 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 2] storage_proxy - Failed to apply mutation from 10.0.3.92#2: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->2 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 6] storage_proxy - Failed to apply mutation from 10.0.3.92#6: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.3.139#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 2] storage_proxy - Failed to apply mutation from 10.0.3.92#2: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->2 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 7] storage_proxy - Failed to apply mutation from 10.0.3.92#7: seastar::semaphore_timed_out (Semaphore timedout)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.3.92#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.3.92#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semaph
ore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
hore)
Nov 09 09:10:47 longevity-mv-si-4d-4-3-db-node-d0cc0297-1 scylla[9901]:  [shard 8] storage_proxy - Failed to apply mutation from 10.0.2.163#8: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#3 13->8 semap
```

**Comment by asias** [comment]
@avikivity Isn't this a scheduling issue? How come the statement cpu time drops so much and the compaction and stream dominate the cpu time during 2011-9-11 08:47 to 09:10 where the QPS dropped to ~zero. 

http://54.171.83.95:3000/d/detailed-master/detailed?orgId=1&from=1604908035000&to=1604915235000

![2020-11-17_11-31](https://user-images.githubusercontent.com/320076/99343287-85e6d580-28c8-11eb-8ef1-2e8fd3427aa0.png)


![image](https://user-images.githubusercontent.com/320076/99343563-1e7d5580-28c9-11eb-87c3-afaf929e066a.png)

**Comment by psarna** [comment]
@asias as for broken promises, it's probably the same cause as https://github.com/scylladb/scylla/issues/7572 (which is not discovered yet)

**Comment by psarna** [comment]
And ignored futures themselves are simply a result of the fact that `mutate_MV` sometimes puts updates in the background, with nobody waiting for them. If such an update fails, its future is inherently ignored, which causes the warning to be printed. Still, these broken promises shouldn't happen and we haven't found their root cause yet. It looks like it's always related to destination nodes being down, but I haven't pinpointed anything.

**Comment by ShlomiBalalis** [comment]
@asias Sorry, I forgot to add the log table to the issue. It is as follows:
```
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|                                                                                           Log links for testrun with test id d0cc0297-f624-4948-8d6f-e9dd9737924d                                                                                           |
+-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Date            | Log type    | Link                                                                                                                                                                                                                        |
+-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 20190101_010101 | prometheus  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/prometheus_snapshot_20201112_224947.tar.gz                                                                                              |
| 20201109_121442 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/db-cluster-d0cc0297.zip                                                                                                 |
| 20201109_121442 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/loader-set-d0cc0297.zip                                                                                                 |
| 20201109_121442 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/monitor-set-d0cc0297.zip                                                                                                |
| 20201110_115037 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/db-cluster-d0cc0297.zip                                                                                                 |
| 20201110_115037 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/loader-set-d0cc0297.zip                                                                                                 |
| 20201110_115037 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/monitor-set-d0cc0297.zip                                                                                                |
| 20201112_231246 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_231246/grafana-screenshot-overview-20201112_231246-longevity-mv-si-4d-4-3-monitor-node-d0cc0297-1.png                          |
| 20201112_231246 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_231246/grafana-screenshot-scylla-per-server-metrics-nemesis-20201112_231559-longevity-mv-si-4d-4-3-monitor-node-d0cc0297-1.png |
| 20201112_232137 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/db-cluster-d0cc0297.zip                                                                                                 |
| 20201112_232137 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/loader-set-d0cc0297.zip                                                                                                 |
| 20201112_232137 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/monitor-set-d0cc0297.zip                                                                                                |
+-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
Notice that there is some repetition in the logs, I'd recommend taking the logs from the latest date if there was no issue with them.
Please let me know if those are enough, I'd like to terminate the nodes if possible.

The monitor can be found in http://54.171.83.95:3000/

**Comment by asias** [comment]
> @asias Sorry, I forgot to add the log table to the issue. It is as follows:
> 
> ```
> +-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> |                                                                                           Log links for testrun with test id d0cc0297-f624-4948-8d6f-e9dd9737924d                                                                                           |
> +-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | Date            | Log type    | Link                                                                                                                                                                                                                        |
> +-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | 20190101_010101 | prometheus  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/prometheus_snapshot_20201112_224947.tar.gz                                                                                              |
> | 20201109_121442 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/db-cluster-d0cc0297.zip                                                                                                 |
> | 20201109_121442 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/loader-set-d0cc0297.zip                                                                                                 |
> | 20201109_121442 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/monitor-set-d0cc0297.zip                                                                                                |
> | 20201110_115037 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/db-cluster-d0cc0297.zip                                                                                                 |
> | 20201110_115037 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/loader-set-d0cc0297.zip                                                                                                 |
> | 20201110_115037 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/monitor-set-d0cc0297.zip                                                                                                |
> | 20201112_231246 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_231246/grafana-screenshot-overview-20201112_231246-longevity-mv-si-4d-4-3-monitor-node-d0cc0297-1.png                          |
> | 20201112_231246 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_231246/grafana-screenshot-scylla-per-server-metrics-nemesis-20201112_231559-longevity-mv-si-4d-4-3-monitor-node-d0cc0297-1.png |
> | 20201112_232137 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/db-cluster-d0cc0297.zip                                                                                                 |
> | 20201112_232137 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/loader-set-d0cc0297.zip                                                                                                 |
> | 20201112_232137 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/monitor-set-d0cc0297.zip                                                                                                |
> +-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> Notice that there is some repetition in the logs, I'd recommend taking the logs from the latest date if there was no issue with them.
> Please let me know if those are enough, I'd like to terminate the nodes if possible.
> 
> The monitor can be found in http://54.171.83.95:3000/

Thanks. I managed to log in one of the nodes to see the logs. Please keep the monitor alive. You can kill the rest of the cluster.

**Comment by ShlomiBalalis** [comment]
> > @asias Sorry, I forgot to add the log table to the issue. It is as follows:
> > ```
> > +-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> > |                                                                                           Log links for testrun with test id d0cc0297-f624-4948-8d6f-e9dd9737924d                                                                                           |
> > +-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> > | Date            | Log type    | Link                                                                                                                                                                                                                        |
> > +-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> > | 20190101_010101 | prometheus  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/prometheus_snapshot_20201112_224947.tar.gz                                                                                              |
> > | 20201109_121442 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/db-cluster-d0cc0297.zip                                                                                                 |
> > | 20201109_121442 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/loader-set-d0cc0297.zip                                                                                                 |
> > | 20201109_121442 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201109_121442/monitor-set-d0cc0297.zip                                                                                                |
> > | 20201110_115037 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/db-cluster-d0cc0297.zip                                                                                                 |
> > | 20201110_115037 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/loader-set-d0cc0297.zip                                                                                                 |
> > | 20201110_115037 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201110_115037/monitor-set-d0cc0297.zip                                                                                                |
> > | 20201112_231246 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_231246/grafana-screenshot-overview-20201112_231246-longevity-mv-si-4d-4-3-monitor-node-d0cc0297-1.png                          |
> > | 20201112_231246 | grafana     | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_231246/grafana-screenshot-scylla-per-server-metrics-nemesis-20201112_231559-longevity-mv-si-4d-4-3-monitor-node-d0cc0297-1.png |
> > | 20201112_232137 | db-cluster  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/db-cluster-d0cc0297.zip                                                                                                 |
> > | 20201112_232137 | loader-set  | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/loader-set-d0cc0297.zip                                                                                                 |
> > | 20201112_232137 | monitor-set | https://cloudius-jenkins-test.s3.amazonaws.com/d0cc0297-f624-4948-8d6f-e9dd9737924d/20201112_232137/monitor-set-d0cc0297.zip                                                                                                |
> > +-----------------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> > ```
> > 
> > 
> > Notice that there is some repetition in the logs, I'd recommend taking the logs from the latest date if there was no issue with them.
> > Please let me know if those are enough, I'd like to terminate the nodes if possible.
> > The monitor can be found in http://54.171.83.95:3000/
> 
> Thanks. I managed to log in one of the nodes to see the logs. Please keep the monitor alive. You can kill the rest of the cluster.

Thank you

**Comment by avikivity** [comment]
> @avikivity Isn't this a scheduling issue? How come the statement cpu time drops so much and the compaction and stream dominate the cpu time during 2011-9-11 08:47 to 09:10 where the QPS dropped to ~zero.
> 
> http://54.171.83.95:3000/d/detailed-master/detailed?orgId=1&from=1604908035000&to=1604915235000
> 
> ![2020-11-17_11-31](https://user-images.githubusercontent.com/320076/99343287-85e6d580-28c8-11eb-8ef1-2e8fd3427aa0.png)
> 

If statement doesn't consume cpu, compaction will eat it up. Of course, it can be a scheduler bug, but the first assumption should be that statements stopped consuming cpu.

Statements can stop consuming cpu if all their concurrency is spent on requests that don't complete (because they depend on a dead node).

**Comment by asias** [comment]
> And ignored futures themselves are simply a result of the fact that `mutate_MV` sometimes puts updates in the background, with nobody waiting for them. If such an update fails, its future is inherently ignored, which causes the warning to be printed. Still, these broken promises shouldn't happen and we haven't found their root cause yet. It looks like it's always related to destination nodes being down, but I haven't pinpointed anything.

Can we discard the future to avoid the massive logs?

**Comment by asias** [comment]
> > @avikivity Isn't this a scheduling issue? How come the statement cpu time drops so much and the compaction and stream dominate the cpu time during 2011-9-11 08:47 to 09:10 where the QPS dropped to ~zero.
> > http://54.171.83.95:3000/d/detailed-master/detailed?orgId=1&from=1604908035000&to=1604915235000
> > ![2020-11-17_11-31](https://user-images.githubusercontent.com/320076/99343287-85e6d580-28c8-11eb-8ef1-2e8fd3427aa0.png)
> 
> If statement doesn't consume cpu, compaction will eat it up. Of course, it can be a scheduler bug, but the first assumption should be that statements stopped consuming cpu.
> 
> Statements can stop consuming cpu if all their concurrency is spent on requests that don't complete (because they depend on a dead node).

One of the node was shutdown,  all live nodes learn the node is down immediately. Shouldn't we fail the corresponding requests on the dead node immediately.  @avikivity , could you please or ask someone who is familiar with the code to check the relevant code?  This is a very serious issue we see a lot of similar issues on lots of tests recently. 

@slivne FYI.

**Comment by avikivity** [comment]
@gleb-cloudius is the expert

**Comment by bhalevy** [comment]
ping @gleb-cloudius

**Comment by slivne** [comment]
We are running repair prior to remove node, if the repaired table is a base for a materialize view and it will find delta - it will place the delta into staging directory (@psarna / @asias please confirm) - once the data is inside staging directory MV will try to build the changes for the MV related to the delta in the base - this is a heavy read operation reading information from partial sstableset (only the current sstables not with staging) and reading from staging only and computing the diff to find the MV update. 

I think we can check the MV metrics in this run to check if during th repair phase we had MV update 
```
The repairs run between 2020-11-09 08:48:18,324 and 2020-11-09 08:57:16,816, and in that time the C-S latency become extremely high (to the point of several seconds) and as a result the number of request the cluster is able to handle nearly reaches zero, as you can see in the images above
```

If there is no place send the MV update it will be recorded as MV hints (@psarna ?) - this can be checked as well

Once the removenode operation starts - its not clear to me if MV / regular hints will be sent - @psarna / @asias  we can check 

All of this is a long shot - yet the spike in CS latency in this specific scenario  related to MV - that was not opened for other cases - seems to indicate it has todo with MV overhead - which I tried to point to in the above.

**Comment by slivne** [comment]
@roydahan - just to double check the removenode procedure "bad behaviour" is confined to MV do we have cases without MV that the removenode operation was run in this manner - did we have any issues in that case ?

**Comment by roydahan** [comment]
I don't have enough examples with big datasets but from what I see it didn't happen in tests where we don't have MV.

**Comment by slivne** [comment]
Thanks roy - @psarna can you please check the above in the "MV expert" hat (https://github.com/scylladb/scylla/issues/7580#issuecomment-739583986)

**Comment by slivne** [comment]
(please note the removenode - is a new nemesis - and the repair before removenode was added based on the docs)

**Comment by psarna** [comment]
@slivne yes, a separate hint manager is always active for materialized views, and if a view update can't be sent, a hint is stored.
There are metrics available for this manager, perhaps you could add tracing these in the scenario?
```bash
[sarna@localhost scylla-ccm]$ curl -s localhost:9180/metrics | grep view | grep hint | grep -v '^#'
scylla_hints_for_views_manager_corrupted_files{shard="0"} 0
scylla_hints_for_views_manager_discarded{shard="0"} 0
scylla_hints_for_views_manager_dropped{shard="0"} 0
scylla_hints_for_views_manager_errors{shard="0"} 0
scylla_hints_for_views_manager_pending_drains{shard="0"} 0.000000
scylla_hints_for_views_manager_pending_sends{shard="0"} 0.000000
scylla_hints_for_views_manager_sent{shard="0"} 0
scylla_hints_for_views_manager_size_of_hints_in_progress{shard="0"} 0.000000
scylla_hints_for_views_manager_written{shard="0"} 0
```

**Comment by slivne** [comment]
Hints without views

![7_1](https://user-images.githubusercontent.com/3465480/102024090-7c9e3b00-3d98-11eb-943a-5d5863279bb4.png)

Hints with views

![7_4](https://user-images.githubusercontent.com/3465480/102024093-7e67fe80-3d98-11eb-83bb-06d4b1581255.png)

So hints are accumulated but are not replayed during the 09:00 --> 09:10 period with no traffic, they are replayed later in high rate and that may be an issue after 09:10 ...

(I expect that when we see the hints being sent is the point in which removenode ends - yet can't really prove it) - I don't see another reason why 

@psarna - I did search for an indication that the view background process when moving sstables from staging is being activated - or has finished but can't find an indication this is the case.

Can you please check which scheduling group this is supposed to run under.

- If this is supposed to run under streaming then it may have ended since between 09:00 --> 09:10 we can't see its running.
- if it doesn't have a scheduling group associated its probably running under main - and this will allow with the large task quota violation we are seeing under main.

@psarna  do we have any metrics for this process to see when it starts and when it ends ?

@psarna since you are the expert (now) - maybe its backpressure overload bad docmputation - we have hints regular and hints for views accumulated - do we somehow consider the amount of hints and will not receive more requests (it would be easy to validate this if the removenode operation would not run immediately after the repairs). Maybe there are metrics you can validate this  - Avi's comment about compaction/main taking more time and not allowing statement to run - maybe correct if we are not reading CQL requests the client has sent because of some backpressure mechanism taking place. 

If we can't figure whats going on - we need to add logs / run this outside of SCT and see if we can figure this out. 

@roydahan - how easy is it to reproduce this - does it reproduce ion 20 minutes as the first nemesis of a run - if so it would be simple to make modifications (without you) to the flow and do further analysis.

**Comment by psarna** [comment]
> 
> @psarna - I did search for an indication that the view background process when moving sstables from staging is being activated - or has finished but can't find an indication this is the case.
> 
> Can you please check which scheduling group this is supposed to run under.

This whole process runs under streaming scheduling group.

> 
>     * If this is supposed to run under streaming then it may have ended since between 09:00 --> 09:10 we can't see its running.
> 
>     * if it doesn't have a scheduling group associated its probably running under main - and this will allow with the large task quota violation we are seeing under main.
> 
> 
> @psarna do we have any metrics for this process to see when it starts and when it ends ?

The following metrics are available:
```cpp
sm::make_gauge("pending_registrations", sm::description("Number of tasks waiting to register staging sstables"),
        [this] { return _registration_sem.waiters(); }),

sm::make_gauge("queued_batches_count", 
        sm::description("Number of sets of sstables queued for view update generation"),
        [this] { return _sstables_with_tables.size(); }),

sm::make_gauge("sstables_to_move_count",
        sm::description("Number of sets of sstables which are already processed and wait to be moved from their staging directory"),
        [this] { return _sstables_to_move.size(); })
```

> 
> @psarna since you are the expert (now) - maybe its backpressure overload bad computation - we have hints regular and hints for views accumulated - do we somehow consider the amount of hints and will not receive more requests (it would be easy to validate this if the removenode operation would not run immediately after the repairs). Maybe there are metrics you can validate this - Avi's comment about compaction/main taking more time and not allowing statement to run - maybe correct if we are not reading CQL requests the client has sent because of some backpressure mechanism taking place.

Hints for views are not taken into account the same way as regular hints - an OverloadedException will only be returned if there are too many regular in-flight hints. If we suspect too many view hints, this metrics can be examined: `scylla_hints_for_views_manager_dropped`.
 
> 
> If we can't figure whats going on - we need to add logs / run this outside of SCT and see if we can figure this out.
>

**Comment by roydahan** [comment]
> Hints without views
> 
> ![7_1](https://user-images.githubusercontent.com/3465480/102024090-7c9e3b00-3d98-11eb-943a-5d5863279bb4.png)
> 
> Hints with views
> 
> ![7_4](https://user-images.githubusercontent.com/3465480/102024093-7e67fe80-3d98-11eb-83bb-06d4b1581255.png)
> 
> So hints are accumulated but are not replayed during the 09:00 --> 09:10 period with no traffic, they are replayed later in high rate and that may be an issue after 09:10 ...
> 
> (I expect that when we see the hints being sent is the point in which removenode ends - yet can't really prove it) - I don't see another reason why
> 
> @psarna - I did search for an indication that the view background process when moving sstables from staging is being activated - or has finished but can't find an indication this is the case.
> 
> Can you please check which scheduling group this is supposed to run under.
> 
> * If this is supposed to run under streaming then it may have ended since between 09:00 --> 09:10 we can't see its running.
> * if it doesn't have a scheduling group associated its probably running under main - and this will allow with the large task quota violation we are seeing under main.
> 
> @psarna do we have any metrics for this process to see when it starts and when it ends ?
> 
> @psarna since you are the expert (now) - maybe its backpressure overload bad computation - we have hints regular and hints for views accumulated - do we somehow consider the amount of hints and will not receive more requests (it would be easy to validate this if the removenode operation would not run immediately after the repairs). Maybe there are metrics you can validate this - Avi's comment about compaction/main taking more time and not allowing statement to run - maybe correct if we are not reading CQL requests the client has sent because of some backpressure mechanism taking place.
> 
> If we can't figure whats going on - we need to add logs / run this outside of SCT and see if we can figure this out.
> 
> @roydahan - how easy is it to reproduce this - does it reproduce ion 20 minutes as the first nemesis of a run - if so it would be simple to make modifications (without you) to the flow and do further analysis.

I think it should reproduce easily.
If you need to reproduce, please ping @ShlomiBalalis in this issue and he will run a reproducer.

**Comment by slivne** [comment]
@ShlomiBalalis - can you provide instructions on how to reproduce the scenario using SCT - and pointers to code so we can do changes and test

**Comment by ShlomiBalalis** [comment]
@slivne Reproducer running - https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/Shlomo/job/longevity-mv-si-4days-test-TerminateAndRemoveNodeMonkey/

**Comment by slivne** [comment]
@psarna took over this and will follow up - @ShlomiBalalis - please update about the status of the run - and if it reproduced - so we know we have a reproducer and then can try and take over, thanks.

**Comment by ShlomiBalalis** [comment]
Scylla version (or git commit hash): 4.3.rc2-0.20201214.ecfe466e7 with build-id 25b5fc3168334b124897a3d9ed178ce02abaeed0
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0427d6c76a7a77e28 (eu-north-1)

On the latest ami, the issue doesn't seem to occur again. There are no write or read timeouts. In their place, there are only repair failures:
```
2020-12-17T11:41:07+00:00  longevity-mv-si-4d-Terminat-db-node-740ef582-3 !WARNING | scylla: [shard 4] repair - repair id [id=1, uuid=5a690d4c-f362-4814-9f02-373ec1834f5d] on shard 4 failed: std::runtime_error (repair id [id=1, uuid=5a690d4c-f362-4814-9f02-373ec1834f5d] on shard 4 failed to repair 132 sub ranges)
2020-12-17 11:41:07.000: (DatabaseLogEvent Severity.ERROR): type=RUNTIME_ERROR regex=std::runtime_error line_number=11517 node=Node longevity-mv-si-4d-Terminat-db-node-740ef582-3 [13.48.29.117 | 10.0.3.73] (seed: False)
```
I don't suppose the failures could be related to this issue? If not, I'll open a new issue for them.

Node list:
```
longevity-mv-si-4d-Terminat-db-node-740ef582-1 [13.49.224.67 | 10.0.3.67] 
longevity-mv-si-4d-Terminat-db-node-740ef582-2 [13.48.70.124 | 10.0.3.102]
longevity-mv-si-4d-Terminat-db-node-740ef582-3 [13.48.29.117 | 10.0.3.73]
longevity-mv-si-4d-Terminat-db-node-740ef582-4 [13.49.80.154 | 10.0.0.187]
longevity-mv-si-4d-Terminat-db-node-740ef582-5 [13.48.26.156 | 10.0.0.245]
longevity-mv-si-4d-Terminat-db-node-740ef582-6 [13.53.198.71 | 10.0.3.203]
longevity-mv-si-4d-Terminat-db-node-740ef582-7 [13.48.105.82 | 10.0.0.185]
longevity-mv-si-4d-Terminat-db-node-740ef582-8 [13.53.96.220 | 10.0.2.166]
longevity-mv-si-4d-Terminat-db-node-740ef582-9 [13.53.78.29 | 10.0.1.34]
longevity-mv-si-4d-Terminat-db-node-740ef582-10 [13.48.105.243 | 10.0.1.172]
longevity-mv-si-4d-Terminat-db-node-740ef582-11 [13.48.178.230 | 10.0.2.152]
longevity-mv-si-4d-Terminat-db-node-740ef582-12 [13.49.226.226 | 10.0.1.58]
```
Monitor: http://13.53.36.9:3000/d/wWHwe_1Mz/scylla-per-server-metrics-nemesis-master?orgId=1&from=now-6h&to=now&refresh=30s

**Comment by bhalevy** [comment]
> Scylla version (or git commit hash): 4.3.rc2-0.20201214.ecfe466e7 with build-id 25b5fc3168334b124897a3d9ed178ce02abaeed0
> Cluster size: 5
> OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0427d6c76a7a77e28 (eu-north-1)
> 
> On the latest ami, the issue doesn't seem to occur again. There are no write or read timeouts. In their place, there are only repair failures:
> 
> ```
> 2020-12-17T11:41:07+00:00  longevity-mv-si-4d-Terminat-db-node-740ef582-3 !WARNING | scylla: [shard 4] repair - repair id [id=1, uuid=5a690d4c-f362-4814-9f02-373ec1834f5d] on shard 4 failed: std::runtime_error (repair id [id=1, uuid=5a690d4c-f362-4814-9f02-373ec1834f5d] on shard 4 failed to repair 132 sub ranges)
> 2020-12-17 11:41:07.000: (DatabaseLogEvent Severity.ERROR): type=RUNTIME_ERROR regex=std::runtime_error line_number=11517 node=Node longevity-mv-si-4d-Terminat-db-node-740ef582-3 [13.48.29.117 | 10.0.3.73] (seed: False)
> ```
> 
> I don't suppose the failures could be related to this issue? If not, I'll open a new issue for them.

What caused this repair?
Was a node killed or spontaneously die during the repair?

/cc @asias

**Comment by fgelcer** [comment]
> > Scylla version (or git commit hash): 4.3.rc2-0.20201214.ecfe466e7 with build-id 25b5fc3168334b124897a3d9ed178ce02abaeed0
> > Cluster size: 5
> > OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0427d6c76a7a77e28 (eu-north-1)
> > On the latest ami, the issue doesn't seem to occur again. There are no write or read timeouts. In their place, there are only repair failures:
> > ```
> > 2020-12-17T11:41:07+00:00  longevity-mv-si-4d-Terminat-db-node-740ef582-3 !WARNING | scylla: [shard 4] repair - repair id [id=1, uuid=5a690d4c-f362-4814-9f02-373ec1834f5d] on shard 4 failed: std::runtime_error (repair id [id=1, uuid=5a690d4c-f362-4814-9f02-373ec1834f5d] on shard 4 failed to repair 132 sub ranges)
> > 2020-12-17 11:41:07.000: (DatabaseLogEvent Severity.ERROR): type=RUNTIME_ERROR regex=std::runtime_error line_number=11517 node=Node longevity-mv-si-4d-Terminat-db-node-740ef582-3 [13.48.29.117 | 10.0.3.73] (seed: False)
> > ```
> > 
> > 
> > I don't suppose the failures could be related to this issue? If not, I'll open a new issue for them.
> 
> What caused this repair?
> Was a node killed or spontaneously die during the repair?
> 
> /cc @asias

IIUC this reproducer is running a nemesis called terminate and remove node, so in the procedure of it, there is a repair, after the node is terminated, and before running `nodetool removenode`. if that is the case, then the issue will be #7806

**Comment by ShlomiBalalis** [comment]
@slivne It does not seem to be reproduced. Should I try to run the reproducer with the same ami I used when I first experienced the error?

**Comment by slivne** [comment]
@ShlomiBalalis
 - can you share the screen shot of what happend when you tried to reproduce with 4.3 current ami
 - yes please can you please try and recreate the scenario on the original ami

**Comment by ShlomiBalalis** [comment]
With the current ami:
![Screenshot from 2020-12-20 18-55-31](https://user-images.githubusercontent.com/40186893/102719143-fa13ff00-42f4-11eb-85e1-b021ef32bbb7.png)

![Screenshot from 2020-12-20 18-52-49](https://user-images.githubusercontent.com/40186893/102719144-fb452c00-42f4-11eb-925e-ca4afeff3043.png)

The latency seems in the normal stage, though there are still some timeouts. I'm not sure if it is normal behavior for that situation.

Rerun with the ami of the original issue (ami-01cc969208fae7a3a):
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/Shlomo/job/longevity-mv-si-4days-test-TerminateAndRemoveNodeMonkey/2/

**Comment by slivne** [comment]
@ShlomiBalalis ping

**Comment by ShlomiBalalis** [comment]
http://34.241.20.14:3000/d/ToGQy4bGz/scylla-per-server-metrics-nemesis-master?orgId=1&refresh=30s&from=now-6h&to=now

![Screenshot from 2020-12-22 14-27-09](https://user-images.githubusercontent.com/40186893/102888625-d4f0cf00-4461-11eb-91a7-3a2d254ebeaf.png)

It seem it was reproduced successfully, though not to the extreme it was before (~4.0 seconds delay compared to up to ~10 seconds)

**Comment by ShlomiBalalis** [comment]
The old ami (ami-01cc969208fae7a3a - rc0) reproduces the issue, while the newer ami (ami-0427d6c76a7a77e28 - rc3) does not. Ergo, it's seems that the issue was solved on its own. Closing.

**Comment by avikivity** [comment]
It is suspicious that the issue was solved on its own.

tuned was not installed on the "old" ami, so it can't be ecfe466e7b36b88dd2f89210676d416fc1c60d1f.

**Comment by slivne** [comment]
We are suspicious of that as well - shlomo is going to run the longer duration test - not only the reproducer and see if he hits thats as well.

**Comment by ShlomiBalalis** [comment]
The issue reoccurred in the latest run of the mv-si longevity:
*Installation details*
Scylla version (or git commit hash): 4.3.rc3-0.20201220.5bd52e4db with build-id ec3427545525638e1720aab150d5addd4b625131
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0554c4148bc9549b0 (eu-north-1)

The issue occurs during the `ScyllaKillMonkey` nemesis. The nemesis forcefully kills all of the scylla process of the target node, and waits until the node service to start them again.
The issue occurred consistently during the two runs of the nemesis (so far):
At `2020-12-23 19:03:32.900` on node 15:
http://13.49.223.117:3000/d/rSoTVHxGz/scylla-per-server-metrics-nemesis-master?orgId=1&from=1608750185667&to=1608750357843
![Screenshot from 2020-12-24 16-11-49](https://user-images.githubusercontent.com/40186893/103093367-ca217000-4602-11eb-9a62-cb7700f77405.png)

And at `2020-12-24 03:29:39.644` no node 14:
http://13.49.223.117:3000/d/rSoTVHxGz/scylla-per-server-metrics-nemesis-master?orgId=1&from=1608780411567&to=1608780972999
![Screenshot from 2020-12-24 16-13-17](https://user-images.githubusercontent.com/40186893/103093414-f63cf100-4602-11eb-9144-0e94abd1bead.png)


Logs:
```
Logs are currently being collected, and will be added soon.
```

**Comment by slivne** [comment]
Its not the same thing - here we are talking about a ~30 seconds window in which the latency is high (and I expect the throughput drops to zero as well ...

In the previous example above there were 10 minutes of no traffic 

its not the same issue

These is simple reproducer Asias has found

https://github.com/scylladb/scylla/issues/7342#issuecomment-707527027

@ShlomiBalalis - (we need to solve this issue - all traffic dies when node is killed) but please see if you get a long duration of no traffic - that is the reproducer for this specific case

**Comment by ShlomiBalalis** [comment]
@ShlomiBalalis Alright, I'll add it to #7342

**Comment by slivne** [comment]
Shlomo's reproducer was without populating all the data - as such it was "easier for scylla"

Changing the scenario to populate all the data we get

![mv_si_cache](https://user-images.githubusercontent.com/3465480/103473188-6c2d1f00-4d9e-11eb-85c8-95496cc59aaf.png)

Almost all the cache depleted
![mv_si_load](https://user-images.githubusercontent.com/3465480/103473227-fb3a3700-4d9e-11eb-8a4d-9125f5fe415d.png)

The drop is still not as severe as shlomo reported above

**Comment by slivne** [comment]
Doing the same with 4.3.rc3 ... to see if its worse

for reference this is the SCT yaml

```
[shlomi@localhost scylla-cluster-tests]$ cat data_dir/c-s_profile_4mv_5queries.yaml
keyspace: mview

keyspace_definition: |

  CREATE KEYSPACE IF NOT EXISTS mview WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 3};

table: users

table_definition: |

  CREATE TABLE IF NOT EXISTS users (
    userid bigint,
    initials int,
    first_name text,
    last_name text,
    password text,
    email text,
    address text,
    userdata blob,
    last_access timeuuid,
    PRIMARY KEY(userid)
  );

extra_definitions:
  - CREATE MATERIALIZED VIEW mview.users_by_first_name AS SELECT userid, first_name, email FROM mview.users WHERE first_name IS NOT NULL and userid IS NOT NULL PRIMARY KEY (first_name, userid);
  - CREATE MATERIALIZED VIEW mview.users_by_last_name AS SELECT userid, last_name, email FROM mview.users WHERE last_name IS NOT NULL and userid IS NOT NULL PRIMARY KEY (last_name, userid);
  - CREATE MATERIALIZED VIEW mview.users_by_email AS SELECT userid, email FROM mview.users WHERE email IS NOT NULL and userid IS NOT NULL PRIMARY KEY (email, userid);
  - CREATE MATERIALIZED VIEW mview.users_by_password AS SELECT userid, password FROM mview.users WHERE password IS NOT NULL and userid IS NOT NULL PRIMARY KEY (password, userid);

columnspec:
  - name: userid
    population: exp(1..10000000)

  - name: initials
    size: fixed(2)
    population: gaussian(1..20000)

  - name: first_name
    size: fixed(5)

  - name: last_name
    size: fixed(5)

  - name: password
    size: fixed(80) # sha-512

  - name: email
    size: uniform(16..50)

  - name: address
    size: uniform(16..50)

  - name: userdata
    size: fixed(10240)

insert:
  partitions: fixed(1)
  batchtype: UNLOGGED

queries:
  read1:
    cql: select * from mview.users where userid = ?
    fields: samerow
  read2:
    cql: select * from mview.users_by_first_name where first_name = ?
    fields: samerow
  read3:
    cql: select * from mview.users_by_last_name where last_name = ?
    fields: samerow
  read4:
    cql: select * from mview.users_by_email where email = ?
    fields: samerow
  read5:
    cql: select * from mview.users_by_password where password = ?
    fields: samerow
```

**Comment by slivne** [comment]
This is with 4.3rc3 - it seems to be the same:

![mv-si-4 3](https://user-images.githubusercontent.com/3465480/103486756-d5497c80-4e08-11eb-83a0-0e1ef1377567.png)

![4 3-mv-si-memory](https://user-images.githubusercontent.com/3465480/103486755-d37fb900-4e08-11eb-88c2-b39230cbfc4f.png)

I am not able to reproduce the original reported issue of not traffic at all for 10 minutes
